### PR TITLE
WIP Collecting `test*.xsh` files when testing

### PR DESCRIPTION
**Pull request by melund**
Very basic implementation. It runs all `test*.xsh` it finds during tests. Fails if they files raises an exception. 

Any other ideas are welcome.

### CODE REVIEW

**Comment by melund** [comment]
image

So we need to make it extract the `def test*()` functions from `test*.xsh` files and run the separately. Maybe a other PR. 

I guess the nice way is to use the build-in xonsh parser/ast functions to do that.  @scopatz  or @adqm do you have any hints how I should do that. .

File: news/test_xsh.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+**Added:** 
+* Added a local py.test framwork to collect and run `test_*.xsh` files.
```
Comment (scopatz) [inline news/test_xsh.rst]: Newline needed here :)

**Comment by scopatz** [comment]
> I guess the nice way is to use the build-in xonsh parser/ast functions to do that. @scopatz or @adqm do you have any hints how I should do that. .

Shouldn't you just import the xsh file (maybe using importlib), then use `dir()` on the module to get all the names in the file, and filter this to just the functions whose names start or end with test?  

It looks like the way it is done now is that it runs the xsh file as a script, rather than just importing it and returning the tests.

**Comment by melund** [comment]
Yes. Now it just runs the file as a script. What you suggest sounds reasonable, but the `conftest.py` is a Python file. Can I just import a xsh from Python with importlib? It didn't occur to me

**Comment by scopatz** [comment]
> Can I just import a xsh from Python with importlib? It didn't occur to me

You can! As long a xonsh.imphook has been imported first, which normally it is since it part of the amalgamation and it is imported by xonsh.main

**Comment by melund** [comment]
So. I finally found a little time to update this. Maybe @santagada and @laerus would be interested in helping with this. Sorry for leaving this so long. 

It now it own pytest plugin. It runs all `test*` functions in `test_*.xsh` files. It doesn't have all the nice assert re-writing that pytest provide for the python tests. But it is a start. 

It is giving me a bit of problems, though. If I run all tests a few of the `xsh` tests fails. But if I run the `test_xonsh.xsh` by it self, the tests pass. 

```
C:\Users\mel\Documents\xonsh>py.test tests\test_xonsh.xsh
============================= test session starts =============================
platform win32 -- Python 3.5.2, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: C:\Users\mel\Documents\xonsh, inifile: setup.cfg
plugins: xdist-1.14
collected 3 items

tests\test_xonsh.xsh ...

========================== 3 passed in 0.56 seconds ===========================
```

**Comment by laerus** [comment]
I will experiment with this these days @melund, thanks for starting it!!

**Comment by melund** [comment]
No problem. And feel free to PR this branch if you come up extra functionality.

**Comment by laerus** [comment]
yea this works when specifying the `test_xonsh.xsh`
if run for all tests i get:
```
=================================================== FAILURES ===================================================
_________________________________________ xonsh test: test_envionment __________________________________________
xonsh execution failed
  File "/home/laerus/projects/xonsh/tests/test_xonsh.xsh", line 8, in test_envionment
    $USER = 'snail'

_________________________________________ xonsh test: test_xonsh_party _________________________________________
xonsh execution failed
  File "/home/laerus/projects/xonsh/tests/test_xonsh.xsh", line 17, in test_xonsh_party
    out = $(echo @(x + ' ' + y))

============================== 2 failed, 1165 passed, 10 skipped in 14.29 seconds ==============================
Exception ignored in: <bound method Execer.__del__ of <xonsh.execer.Execer object at 0x7f94000d2550>>
Traceback (most recent call last):
  File "/home/laerus/projects/xonsh/xonsh/execer.py", line 46, in __del__
  File "/home/laerus/projects/xonsh/xonsh/built_ins.py", line 751, in unload_builtins
  File "/usr/lib/python3.5/abc.py", line 187, in __instancecheck__
AttributeError: 'NoneType' object has no attribute '_abc_invalidation_counter'
Exception ignored in: <generator object get_tokens at 0x7f93f8560938>
Traceback (most recent call last):
  File "/home/laerus/projects/xonsh/xonsh/lexer.py", line 271, in get_tokens
TypeError: catching classes that do not inherit from BaseException is not allowed
```

**Comment by laerus** [comment]
we've seen some of the exceptions before #1335

**Comment by santagada** [comment]
the last stack trace is non related to the failures and happen on my machine and on travis

**Comment by melund** [comment]
True. I didn't see that stack trace either. In the beginning I implemented it as a local conftest.py plugin. That seemed to work fine. It only started to fail when I moved it into xonsh it self as public pytest plugin.

My idea was that it would be good for xontribs developers to use it as well. But I can move it back to the conftest.py file, so we have something that is useable.

**Comment by santagada** [comment]
I really want to use it, but maybe first merging it as conftest and then we try again to move to a pytest plugin

**Comment by melund** [comment]
@santagada ... I will do that.

**Comment by melund** [comment]
@santagada .. Sorry. I was mistaken. I didn't run before either, and the problem is something else entirely. 

Tests fail if they run after certain other tests. Maybe they don't clean up nicely after them self. 

I haven't checked all our test files, but the following cause problems if they are run before the *.xsh tests. 
```
test_aliases.py
test_contexts.py
test_dirstack.py
test_environ.py
test_events.py
test_execer.py 
test_history.py 
test_main.py
test_parser.py
```
while these don't cause problems:
```
test_ast.py
test_builtins.py
test_foreign_shells.py
test_imphooks.py
test_inspectors.py
test_lazyasd.py 
test_lazyjson.py
test_lexer.py
test_man.py
test_news.py
test_ptk_history.py
``` 

I think a quick workaround is to force pytest to run `test_*.xsh` tests first. 

/Morten

**Comment by melund** [comment]
@santagada and @laerus ... I think it is working now. I had to implement a hack to ensure that `test_*.xsh` tests are run before python tests. I think there is an issue with the pytest not cleaning up correctly.

**Comment by laerus** [comment]
Looks good to me, thanks for this @melund! maybe it's safe to merge as this doesn't affect the rest of the tests.

**Comment by scopatz** [comment]
Go for it @laerus :)

**Comment by scopatz** [comment]
Thank you all so much for this!

**Pull request merged**